opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,154 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=abstract-class-instantiated
+# pylint: disable=too-many-ancestors
+# pylint: disable=useless-super-delegation
+
+# type: ignore
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from re import ASCII
+from re import compile as compile_
+from typing import Callable
+
+_logger = getLogger(__name__)
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}", ASCII)
+
+    @abstractmethod
+    def __init__(self, name, *args, unit="", description="", **kwargs):
+
+        if self._name_regex.fullmatch(name) is None:
+            raise Exception("Invalid instrument name {}".format(name))
+
+        if len(unit) > 63:
+            raise Exception("unit must be 63 characters or shorter")
+
+        if any(ord(character) > 127 for character in unit):
+            raise Exception("unit must only contain ASCII characters")
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+    @abstractmethod
+    def __init__(
+        self, name, callback, *args, unit="", description="", **kwargs
+    ):
+
+        if not isinstance(callback, Callable):
+            raise Exception("callback must be callable")
+
+        super().__init__(
+            name, unit=unit, description=description, *args, **kwargs
+        )
+
+    def observe(self):
+        return next(self._callback)  # pylint: disable=no-member
+
+
+class Adding(Instrument):
+    pass
+
+
+class Grouping(Instrument):
+    pass
+
+
+class Monotonic(Adding):
+    pass
+
+
+class NonMonotonic(Adding):
+    pass
+
+
+class Counter(Monotonic, Synchronous):
+    @abstractmethod
+    def add(self, amount, **attributes):
```

Comment:
We should not have used a dict in Tracer. It is exactly for this kind of situations that the `**` feature was made for. I have made the positional arguments positional only :+1:

Reply:
> We should not have used a dict in Tracer. It is exactly for this kind of situations that the ** feature was made for.

I'd rather the tracing and metrics parts be consistent. I don't agree this is idiomatic usage of named arguments and `**`. I've usually seen this as a shorthand for passing named arguments through multiple functions, higher order functions, decorators, etc. I think `**{"some.invalid.identifier": ...}` is a workaround to the problem.

It also forces a copy of the dictionary every time you use the `**` operator. Is there way we could support both calling conventions maybe?

> I have made the positional arguments positional only

It looks like the `/` for this is only supported in Python 3.8+

Reply:
Sorry can you explain what you mean with this?

> I've usually seen this as a shorthand for passing named arguments through multiple functions, higher order functions, decorators, etc. I think **{"some.invalid.identifier": ...} is a workaround to the problem.

> It also forces a copy of the dictionary every time you use the ** operator. 

Why is this advantageous?

Reply:
> Sorry can you explain what you mean with this?

To me, `*args`/`**kwargs` is best used for functions that need to behave very dynamically, like the `functools` module. We don't have that requirement here, we just take the value and an attribute map.

> Why is this advantageous?

It's not advantageous, I am saying it does a lot of unnecessary work (and metrics are recorded in hot code paths). With this signature, it is impossible to not create a new dictionary every time the function is called.

IIUC, avoiding expensive attribute copying is the purpose of this section of the spec:

> [the client MAY allow attribute values to be passed in using a more efficient way (e.g. strong typed struct allocated on the callstack, tuple)](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#:~:text=the%20counter%20creation%2C-,the%20client%20MAY%20allow%20attribute%20values%20to%20be%20passed%20in%20using%20a%20more%20efficient%20way%20(e.g.%20strong%20typed%20struct%20allocated%20on%20the%20callstack%2C%20tuple),-.%20The%20API%20MUST)

Reply:
Discussed in SIG, we will go with accepting attributes parameter instead of capturing them as named arguments

Reply:
I have changed the code to use a dictionary. I still disagree on not using `**` but because it is not possible to use the `/` argument separator for Python 3.7 then I'll go with this approach.